ConcurrentAPI.adoc

```diff
@@ -0,0 +1,220 @@
+== Distributed Concurrent API
+
+=== Motivation
+
+Coming from an initial issue where we need to provide an implementation for vert.x cluster manager for the shared data feature, provide concurrent and distributed programming capabilities.
+
+* Distributed locks -> is it a good idea ?
+* Distributed semaphores : first use case, vert-x cluster management
+* Distributed Countdown latch : it seems the team has already spoken about having it
+* Distributed Atomic Variables : we already have them, would be worth to offer an specific AtomicLong (for example) ?
+
+=== Scope
+
+All these interfaces should be placed in commons, in order to be used either in Embedded and Server mode.
+
+== Infinispan Distributed Locks
+
+I'm wondering if we need to expose Locks on a API. Not sure, because we can already Lock a key on the Cache API and I believe this could be confusing.
+Not sure we need to add anything else right now.
+
+```java
+
+public interface InfinispanLock extends Lock {
+
+}
+
+```
+
+If we decide to do it, the actual implementation could be a wrapper on JGroups Locks.
+
+== Infinipan Distributed Semaphores
+
+Distributed version of java.util.concurrent.Semaphore.
+
+http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html
+
+To create a Semaphore, use the Infinispan Utility class, passing the semaphoreConfig or using the default values or XML values.
+Configuration can be ovrerriden programatically.
+
+First Semaphore wont't implement fairness, this can be cosidered after.
```

Comment:
Ok. Anyway, we are not going to start with a semaphore yet, but a distributed async lock that could be released by any thread. So like a semaphore with 1 permit.